xrpl/utils/tx_parser/utils.py

```diff
@@ -0,0 +1,850 @@
+        )
+
+    # Else it has modified an offer.
+    # Status is 'partially-filled' or 'filled'.
+    final_amount = (
+        _parse_currency_amount(getattr(node.final_fields, side))
+        if (hasattr(node.final_fields, side))
+        else None
+    )
+    previous_amount = (
+        _parse_currency_amount(getattr(node.previous_fields, side))
+        if (hasattr(node.previous_fields, side))
+        else None
+    )
+    value = _calculate_delta(final_amount, previous_amount)
+
+    change_amount = ChangeAmount(
+        final_amount=final_amount.__dict__, previous_value=str(0 - value)
+    )
+
+    return change_amount
+
+
+def _get_quality(node: NormalizedNode) -> str:
+    """Calculate the offers quality.
+
+    Args:
+        node (NormalizedNode): Normalized node.
+
+    Returns:
+        str: The offers quality.
+    """
+    taker_gets = (
+        node.final_fields.TakerGets
+        if (hasattr(node.final_fields, "TakerGets"))
+        else node.new_fields.TakerGets
+    )
+    taker_pays = (
+        node.final_fields.TakerPays
+        if (hasattr(node.final_fields, "TakerPays"))
+        else node.new_fields.TakerPays
+    )
+    taker_gets_value = (
+        drops_to_xrp(taker_gets) if (isinstance(taker_gets, str)) else taker_gets.value
+    )
+    taker_pays_value = (
+        drops_to_xrp(taker_pays) if (isinstance(taker_pays, str)) else taker_pays.value
+    )
+
+    if Decimal(taker_gets_value) > 0 and Decimal(taker_pays_value) > 0:
+        return str(Decimal(taker_pays_value) / Decimal(taker_gets_value))
+
+    return "0"
+
+
+def _ripple_to_unix_timestamp(rpepoch: int) -> int:
+    return rpepoch + 0x386D4380
+
+
+def _get_expiration_time(node: NormalizedNode) -> Union[Any, None]:
+    """Formats the ripple timestamp to a easy to read format.
+
+    Args:
+        node (NormalizedNode): Normalized node.
+
+    Returns:
+        Union[Any, None]:
+            Expiration time in a easy to read format.
+    """
+    expiration_time = (
+        node.final_fields.Expiration
+        if (hasattr(node.final_fields, "Expiration"))
+        else node.new_fields.Expiration
+        if (hasattr(node.new_fields, "Expiration"))
+        else None
+    )
+    if not isinstance(expiration_time, int):
+        return expiration_time
+
+    return str(
+        datetime.utcfromtimestamp(_ripple_to_unix_timestamp(expiration_time)).strftime(
+            "%Y-%m-%d %H:%M:%S"
+        )
+    )
+
+
+def _remove_undefined(order: OrderChange) -> OrderChange:
+    """Remove all attributes that are 'None'.
+
+    Args:
+        order (OrderChange): Order change.
+
+    Returns:
+        OrderChange:
+            Cleaned up OrderChange object.
+    """
+    order_dict = order.__dict__.copy()
+
+    for k, v in order_dict.items():
+        if v is None:
+            delattr(order, k)
+
+    return order
+
+
+def _calculate_received_and_paid_amount(
+    taker_gets: Any,
+    taker_pays: Any,
+    direction: str,
+) -> Any:
```

Comment:
When you return multiple things you can represent that with `Tuple[first_type, second_type]`